stylo: Support linear-gradients as background-image

emilio opened the pull request:
This PR supports setting the background-image property to a linear-gradient in Geckolib.

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes do not require tests because these are geckolib-only changes.

r? @mbrubeck 
cc @heycam @bholley

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/11456)

<!-- Reviewable:end -->

bholley committed:
Stop mapping overflow-wrap to mWordWrap.

This was fixed in gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=955857

emilio committed:
stylo: Support linear-gradients as background-image

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/values.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style code, but no tests are modified. Please consider adding a test!

emilio commented:
Gecko bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1276085

mbrubeck commented on ports/geckolib/properties.mako.rs:
```diff
@@ -700,10 +705,134 @@ fn static_assert() {
+                        // NB: stops are guaranteed to be none in the gecko side by
+                        // default.
+                        stop.position.to_gecko_style_coord(&mut coord.mUnit,
+                                                           &mut coord.mValue);
```
`coord.set(stop.position)`

mbrubeck commented on ports/geckolib/properties.mako.rs:
```diff
@@ -700,10 +705,134 @@ fn static_assert() {
+                    if let AngleOrCorner::Angle(angle) = gradient.angle_or_corner {
+                        unsafe {
+                            angle.to_gecko_style_coord(&mut (*gecko_gradient).mAngle.mUnit,
+                                                       &mut (*gecko_gradient).mAngle.mValue);
```
You can write this as `gecko_gradient.mAngle.set(angle);`.

mbrubeck commented:
r=mbrubeck with some minor cleanup suggestions (above).

highfive commented:
New code was committed to pull request.

emilio commented:
@mbrubeck: Thanks for the fast review! Forgot about those methods :)

Addressed the comments, maybe worth waiting for @heycam's review in Gecko to see if any other change is required.

emilio commented:
Gecko changes got r+'d, so I'm going to land this (per the `r=mbrubeck` above) and make the nsTArray bindings work in a separate PR today.

@bors-servo: r=mbrubeck

bors-servo commented:
:pushpin: Commit bd72743 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck bd7274307114fbc29a75cb5d8697c1c92a6ae275 -->

bors-servo commented:
:hourglass: Testing commit bd72743 with merge e3e1cc8...

emilio committed:
stylo: Allow setting nsTArray members directly

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1336)

highfive commented:
▶ TIMEOUT [expected PASS] /css-transforms-1_dev/html/svg-translate-multiple-002.htm

emilio commented:
@bors-servo:  retry

 * Possibly new intermittent?

highfive commented:
New code was committed to pull request.

emilio committed:
stylo: Refactor nsTArray sugar and remove unused import

highfive commented:
New code was committed to pull request.

emilio commented:
Added the nsTArray bindings, had to add a patch from @bholley that fixes overflow-wrap with the new bindings. cc @heycam about that.

Note that this won't build until #11445 lands, since it uses `copy_from()`.

re-r? @mbrubeck

highfive commented:
New code was committed to pull request.

emilio committed:
stylo: Remove unneeded null-check, and isolate unsafe code

Gecko's allocs are infallible by default.

highfive commented:
New code was committed to pull request.

mbrubeck commented:
-S-blocked-on-external -S-awaiting-review +S-awaiting-answer 

Looks good to me, with one suggested change. r=mbrubeck with or without the change.

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 5 of 5 files at r1, 1 of 1 files at r2, 6 of 6 files at r3, 2 of 9 files at r6, 4 of 6 files at r7, 1 of 1 files at r8, 1 of 2 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[ports/geckolib/gecko_bindings/sugar/ns_t_array.rs, line 9 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11456#-KJGom1mh1FZWq75fwrY:-KJGom1n2fuDxlHXT1L6:1164355462) ([raw file](https://github.com/servo/servo/blob/0d9455ec1399299ae24e2914b98da71ec1751a2d/ports/geckolib/gecko_bindings/sugar/ns_t_array.rs#L9)):*
> ```Rust
> use structs::{nsTArray, nsTArrayHeader};
> 
> impl<T> Index<u32> for nsTArray<T> {
> ```

Would it make sense to instead impl Deref/DerefMut with `Target=[T]` for `nsTArray<T>`? Then we would get `len`, `Index`, `IndexMut` and other slice methods for free.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11456#-:-KJGpiViLrGbhd13h_cg:-1357008967)*
<!-- Sent from Reviewable.io -->

emilio committed:
stylo: Use Deref<Target=[T]> for nsTArray instead of Index<u32>

Makes things a lot nicer :)

highfive commented:
New code was committed to pull request.

emilio commented:
Yep, that's a great idea, I don't know why I didn't thought about it :)

@bors-servo: r=mbrubeck

bors-servo commented:
:pushpin: Commit c6340c4 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck c6340c4840fb46cf6f9a8fc5deaa2ff1095901d9 -->

bors-servo commented:
:hourglass: Testing commit c6340c4 with merge e23962c...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1250), [arm32](http://build.servo.org/builders/arm32/builds/1331), [arm64](http://build.servo.org/builders/arm64/builds/1317), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1339), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1397), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1381), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1365), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1417), [windows](http://build.servo.org/builders/windows/builds/1048)
